humility-core/src/hubris.rs

```diff
@@ -3034,6 +3034,38 @@ impl HubrisArchive {
         !self.modules.is_empty()
     }
 
+    ///
+    /// Takes a list of potentially similar types and deduplicates the list.
+    ///
+    fn dedup(&self, goffs: Vec<HubrisGoff>) -> Result<Vec<HubrisGoff>> {
```

Comment:
Looking at how this is used, we could save a few allocations by instead taking an `Iterator<Item = &HubrisGoff>`.

This would also require modifying call sites; here's a patch:
```diff
diff --git a/humility-core/src/hubris.rs b/humility-core/src/hubris.rs
index dd807c0..7a2f143 100644
--- a/humility-core/src/hubris.rs
+++ b/humility-core/src/hubris.rs
@@ -3037,33 +3037,32 @@ impl HubrisArchive {
     ///
     /// Takes a list of potentially similar types and deduplicates the list.
     ///
-    fn dedup(&self, goffs: Vec<HubrisGoff>) -> Result<Vec<HubrisGoff>> {
+    fn dedup<'a, I: Iterator<Item = &'a HubrisGoff>>(
+        &self,
+        goffs: I,
+    ) -> Result<Vec<HubrisGoff>> {
         let mut cmp = Vec::new();
         let mut out = Vec::new();
 
-        if goffs.len() <= 1 {
-            Ok(goffs)
-        } else {
-            for goff in goffs {
-                cmp.push((goff, self.lookup_type(goff)?));
-            }
+        for &goff in goffs {
+            cmp.push((goff, self.lookup_type(goff)?));
+        }
 
-            for i in 0..cmp.len() {
-                let mut dup = false;
+        for i in 0..cmp.len() {
+            let mut dup = false;
 
-                for j in i + 1..cmp.len() {
-                    if !cmp[i].1.differs(self, &cmp[j].1)? {
-                        dup = true;
-                        break;
-                    }
+            for j in i + 1..cmp.len() {
+                if !cmp[i].1.differs(self, &cmp[j].1)? {
+                    dup = true;
+                    break;
                 }
+            }
 
-                if !dup {
-                    out.push(cmp[i].0);
-                }
+            if !dup {
+                out.push(cmp[i].0);
             }
-            Ok(out)
         }
+        Ok(out)
     }
 
     ///
@@ -3075,7 +3074,7 @@ impl HubrisArchive {
     pub fn lookup_struct_byname(&self, name: &str) -> Result<&HubrisStruct> {
         match self.structs_byname.get_vec(name) {
             Some(v) => {
-                let m = self.dedup(v.to_vec())?;
+                let m = self.dedup(v.iter())?;
 
                 if m.len() > 1 {
                     Err(anyhow!("{} matches more than one structure", name))
@@ -5603,12 +5602,8 @@ impl HubrisModule {
     ) -> Result<&'a HubrisStruct> {
         match hubris.structs_byname.get_vec(name) {
             Some(v) => {
-                let m = hubris.dedup(
-                    v.iter()
-                        .filter(|g| g.object == self.object)
-                        .copied()
-                        .collect::<Vec<HubrisGoff>>(),
-                )?;
+                let m = hubris
+                    .dedup(v.iter().filter(|g| g.object == self.object))?;
 
                 if m.len() > 1 {
                     Err(anyhow!("{} matches more than one structure", name))
@@ -5632,12 +5627,8 @@ impl HubrisModule {
     ) -> Result<&'a HubrisEnum> {
         match hubris.enums_byname.get_vec(name) {
             Some(v) => {
-                let m = hubris.dedup(
-                    v.iter()
-                        .filter(|g| g.object == self.object)
-                        .copied()
-                        .collect::<Vec<HubrisGoff>>(),
-                )?;
+                let m = hubris
+                    .dedup(v.iter().filter(|g| g.object == self.object))?;
 
                 if m.len() > 1 {
                     Err(anyhow!("{} matches more than one enum", name))
```

Reply:
Looks good; merged this with your other suggestion on avoiding the indexing, and the result is cleaner. Thank you!